include/onnxruntime/core/framework/tensor.h

```diff
@@ -74,6 +76,7 @@ class Tensor final {
    */
   Tensor(MLDataType p_type, const TensorShape& shape, std::shared_ptr<IAllocator> allocator, int64_t offset = 0);
 
+  void InitTensor(MLDataType p_type, const TensorShape& shape, std::shared_ptr<IAllocator> allocator, int64_t offset = 0);
   ~Tensor();
```

Comment:
Constructing a Tensor and using std::move semantics to update an existing instance would be a bit cleaner.

Reply:
Sure, this would work too. Why would exposing an Init method be any less clean though? You've a default constructed object and an Init method to initialize it.

Reply:
The existing code has move assignment. 

I'd argue that the semantics are clearer to go via a constructor and move assignment. Constructor always starts with a clean slate so it's clear what it needs to handle. InitTensor could be called at any point in time, and multiple times, so potentially you start adding logic (in theory) to take care of cleaning up any existing state to get back to a clean state in order to do the init. e.g. should the init be running the same logic as the destructor to cleanup any buffer that exists at the time?

---
In reply to: [331361908](https://github.com/microsoft/onnxruntime/pull/1983#discussion_r331361908) [](ancestors = 331361908)

Reply:
Ok.